[web-animations] baseline-shift animation is incorrect

graouts opened the pull request:
#### 50e415a29df2b2c9ca9541dedf0cbda6878b1f95
<pre>
[web-animations] baseline-shift animation is incorrect
<a href="https://bugs.webkit.org/show_bug.cgi?id=248178">https://bugs.webkit.org/show_bug.cgi?id=248178</a>

Reviewed by Darin Adler.

We fail the first subtest of the WPT test web-animations/responsive/baselineShift.html because we only
do part of the work required to animate the baseline-shift property and also fail to return the correct
computed value for it.

The baseline-shift property can be a &lt;length-percentage&gt; or a keyword. The &lt;length-percentage&gt; is contained
in SVGRenderStyle::baselineShiftValue() while the keyword is contained in SVGRenderStyle::baselineShift().
We create a dedicated animation wrapper that accounts for both of those methods.

Now that SVGRenderStyle has the correct value for baselineShift() when blending, we must also ensure that
we return a _computed_ value and not just serialize the value held in SVGRenderStyle::baselineShiftValue().
To that end, we change SVGLengthValue::toCSSPrimitiveValue() to convert the stored value using
SVGLengthContext::convertValueToUserUnits() when provided an Element. We also change that method to not be
static since its first parameter is an SVGLengthValue.

* LayoutTests/imported/w3c/web-platform-tests/web-animations/responsive/baselineShift-expected.txt:
* LayoutTests/svg/css/parse-calc-length-expected.txt:
* LayoutTests/svg/css/parse-calc-length.html:
* LayoutTests/svg/css/scientific-numbers-expected.txt:
* LayoutTests/svg/css/scientific-numbers.html:
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):
* Source/WebCore/css/SVGCSSComputedStyleDeclaration.cpp:
(WebCore::strokeDashArrayToCSSValueList):
(WebCore::ComputedStyleExtractor::svgPropertyValue):
* Source/WebCore/svg/SVGLengthValue.cpp:
(WebCore::SVGLengthValue::toCSSPrimitiveValue const):
(WebCore::SVGLengthValue::toCSSPrimitiveValue): Deleted.
* Source/WebCore/svg/SVGLengthValue.h:

Canonical link: <a href="https://commits.webkit.org/256934@main">https://commits.webkit.org/256934@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/035112aa9c1962f3b9609dfa16e55c82832aff13

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/97286 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/6550 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/30436 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/106805 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/167071 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/101255 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/6846 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/35285 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/89680 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/103486 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/102949 "Passed tests") | [⏳ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/iOS-16-Simulator-WK2-Tests-EWS "Waiting to run tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/83914 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/32129 "Passed tests") | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/88817 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/75055 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/566 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/20288 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/550 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/21747 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/5351 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/44234 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2348 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/1794 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/41076 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/animation/CSSPropertyAnimation.cpp:
```diff
@@ -2429,6 +2429,36 @@ class PropertyWrapperFontSizeAdjust final : public PropertyWrapperGetter<std::op
+        if (a.svgStyle().baselineShift() != b.svgStyle().baselineShift())
+            return false;
+
+        return PropertyWrapper::equals(a, b);
```
I suggest using `&&` here and making it a one line function.

darinadler commented on Source/WebCore/svg/SVGLengthValue.h:
```diff
@@ -67,7 +68,7 @@ class SVGLengthValue {
 
     static SVGLengthValue fromCSSPrimitiveValue(const CSSPrimitiveValue&);
-    static Ref<CSSPrimitiveValue> toCSSPrimitiveValue(const SVGLengthValue&);
+    Ref<CSSPrimitiveValue> toCSSPrimitiveValue(const Element*) const;
```
I suggest `const Element&` instead, unless it can be `nullptr`.

graouts replied:
In the end we do need to pass `nullptr` sometimes, so I kept it as such.

darinadler commented on Source/WebCore/css/SVGCSSComputedStyleDeclaration.cpp:
```diff
@@ -50,14 +50,14 @@ static RefPtr<CSSPrimitiveValue> glyphOrientationToCSSPrimitiveValue(GlyphOrient
 }
 
-static Ref<CSSValue> strokeDashArrayToCSSValueList(const Vector<SVGLengthValue>& dashes)
+static Ref<CSSValue> strokeDashArrayToCSSValueList(const Vector<SVGLengthValue>& dashes, const Element* element)
```
I suggest `const Element&` instead, unless it can be `nullptr`.

graouts committed:
[web-animations] baseline-shift animation is incorrect
https://bugs.webkit.org/show_bug.cgi?id=248178

Reviewed by Darin Adler.

We fail the first subtest of the WPT test web-animations/responsive/baselineShift.html because we only
do part of the work required to animate the baseline-shift property and also fail to return the correct
computed value for it.

The baseline-shift property can be a <length-percentage> or a keyword. The <length-percentage> is contained
in SVGRenderStyle::baselineShiftValue() while the keyword is contained in SVGRenderStyle::baselineShift().
We create a dedicated animation wrapper that accounts for both of those methods.

Now that SVGRenderStyle has the correct value for baselineShift() when blending, we must also ensure that
we return a _computed_ value and not just serialize the value held in SVGRenderStyle::baselineShiftValue().
To that end, we change SVGLengthValue::toCSSPrimitiveValue() to convert the stored value using
SVGLengthContext::convertValueToUserUnits() when provided an Element. We also change that method to not be
static since its first parameter is an SVGLengthValue.

* LayoutTests/imported/w3c/web-platform-tests/web-animations/responsive/baselineShift-expected.txt:
* LayoutTests/svg/css/parse-calc-length-expected.txt:
* LayoutTests/svg/css/parse-calc-length.html:
* LayoutTests/svg/css/scientific-numbers-expected.txt:
* LayoutTests/svg/css/scientific-numbers.html:
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):
* Source/WebCore/css/SVGCSSComputedStyleDeclaration.cpp:
(WebCore::strokeDashArrayToCSSValueList):
(WebCore::ComputedStyleExtractor::svgPropertyValue):
* Source/WebCore/svg/SVGLengthValue.cpp:
(WebCore::SVGLengthValue::toCSSPrimitiveValue const):
(WebCore::SVGLengthValue::toCSSPrimitiveValue): Deleted.
* Source/WebCore/svg/SVGLengthValue.h:

Canonical link: https://commits.webkit.org/256934@main
